slackbot/bot.py

```diff
@@ -54,3 +70,96 @@ def wrapper(func):
+    return wrapper
+
+def schedule(frequency):
+    """
+    Supports scheduling a task with basic grammer.
+
+    For scheduling a one time task:
+        @schedule("in 20 mins")
+        def job()
+            return
+
+    For scheduling a periodic task:
+        @schedule("every 20 minutes")
+        def job()
+            return
+
+    Similarly:
+        @schedule("every 2 hours and 30 minutes")
+        @schedule("every Monday at 10:00 AM")
+        @schedule("every day at 9:00 AM")
+
+    TODO:
+      Support for "on <day> at <time>" format as well.
+    """
+
+    assert py_schedule, "@schedule was called without installing 'schedule'"
+
+    def _schedule_decorator(func):
+        seconds = re.findall("(\d+)\ssecond", frequency, re.IGNORECASE)
+        minutes = re.findall("(\d+)\sminute", frequency, re.IGNORECASE)
+        hours = re.findall("(\d+)\shour", frequency, re.IGNORECASE)
+        days_of_week = re.findall(
+            "monday|tuesday|wednesday|thursday|friday|saturday|sunday",
+            frequency,
+            re.IGNORECASE)
+        every_day = re.findall("\sday\s", frequency, re.IGNORECASE)
+        time_tuple = re.findall("at\s(.*)\s(AM|PM)", frequency, re.IGNORECASE)
+
+        # TODO: Sanity checks - only one of minutes, hours, days_of_week should
+        # be set at a time.
+
+        time_24hr = '00:00'
+        if time_tuple:
+            time_24hr = convert_time("%s %s" % (time_tuple[0][0], time_tuple[0][1]))
+
+        time_in_seconds = 0
+        if seconds:
+            time_in_seconds += int(seconds[0])
+        if minutes:
+            time_in_seconds += int(minutes[0]) * 60
+        if hours:
+            time_in_seconds += int(hours[0]) * 60 * 60
+
+        if frequency.startswith('in'):
+            # Schedule a one time task.
+            @wraps(func)
+            def job(*args, **kwargs):
+                time.sleep(time_in_seconds)
+                return func(*args, **kwargs)
+
+            # Execute this task in a new thread.
+            _thread.start_new_thread(job, tuple())
+            return
+        elif frequency.startswith('every'):
+            # Schedule a periodic task.
+            if time_in_seconds is not 0 :
+                job = py_schedule.every(int(seconds[0])).seconds.do(func)
+                logger.debug('scheduled job: %s' % job)
+                return
+
+            if every_day:
+                job = py_schedule.every().day.at(time_24hr).do(func)
+                logger.debug('scheduled job: %s' % job)
+                return
+
+            if days_of_week:
+                for day_of_week in days_of_week:
+                    every = py_schedule.every()
+                    day_func = getattr(every, day_of_week.lower())
+                    job = day_func.at(time_24hr).do(func)
+                    logger.debug('scheduled job: %s' % job)
+                return
+        elif frequency.startswith('on'):
+            # Schedule a one time task at a specific time.
```

Comment:
Instead of using this natural style language ("every 1 hour and 30 minutes.."), which is more error prone, I prefer to use the DSL provided by the scheduler library, e.g:

```python
@periodic_job(interval=schedule.every().hour.at(":30"))
def do_something():
   ...
```